src/lib/Merkle.nim

```diff
@@ -1,104 +1,183 @@
+    a leaf E. Adding it to the sequence looks like:
+
+        A B C D E
+
+    In order to create the Merkle tree for this, we build
+    it up as we had before, but we'll be missing some elements.
+    We'll fill those in with "?".
+
+                   ABCDE???
+             ABCD           E???
+         AB       CD     E?      ?? 
+       A    B   C    D E    ?  ?    ?
+
+    In practice, these "?"s are represented by `nil`. So (I)
+    then reads:
+
+        All leafs to the left of a non-"?" leaf are non-"?",
+        and all leafs to the right of a non-"?" leaf are "?".
+
+    See that it holds for this example, and also when we add
+    another node F:
+
+                   ABCDEF??
+             ABCD           EF??
+         AB       CD     EF      ?? 
+       A    B   C    D E    F  ?    ?
+
+   And intuitively should keep holding forever.
+
+   Now, how the "?"s are treated is actually slightly
+   subtle. Essentially, when combining two trees,
+   if the right tree is missing (since ONLY the right
+   tree can, due to (I)), we treat it as a duplicate
+   of the left tree. Thus if we denote "copies" with
+   lowercase letters, then
+
+         A?        is treated as        Aa
+       A    ?                         A    a
+
+   And the previous large tree is treated as:
+
+                   ABCDEFef
+             ABCD           EFef
+         AB       CD     EF      ef 
+       A    B   C    D E    F
+
+   See that the bottom-rightmost two leaves are disregarded
+   entirely, since we can get the value of their parent, ef,
+   from its sibling, EF.
+
+]#
+
+# -- Creation -- #
+
+func newLeaf(hash: string): Merkle {.raises: [].} =
+    result = Merkle(isLeaf: true, hash: hash)
+
+proc rehash(tree: Merkle) {.raises: [].}  # Forward declaration for use in next proc
+proc newBranch(left: Merkle, right: Merkle): Merkle {.raises: [].} =
+    result = Merkle(isLeaf: false, hash: "", left: left, right: right)
+    result.rehash()
+
+proc newChainOfDepth(depth: int, hash: string): Merkle {.raises: [].} =
+    ## O(log n) proc to create a Merkle tree populated only by leftmost items,
+    ## terminating in a leaf.
+    ## Thus is used when adding a new hash to a full tree, as illustrated in
+    ## the big-ass comment way above (upon adding leaf E)
+    if depth == 0:
+        return newLeaf(hash)
+    else:
+        return newBranch(newChainOfDepth(depth - 1, hash), nil)
+
+# -- Properties -- #
+
+func isBranch(tree: Merkle): bool {.raises: [].} =
+    return not tree.isLeaf
+
+func isFull(tree: Merkle): bool {.raises: [].} =
+    ## Are all 2^depth leaf nodes populated?
+    if tree.isLeaf:
+        return true
+    if tree.right.isNil:
+        return false
+    return tree.left.isFull and tree.right.isFull
+
+func depth(tree: Merkle): int {.raises: [].} =
+    ## Depth of a Merkle tree. We consider a leaf to have depth 0, so the empty tree has depth -1
+    if tree.isLeaf:
+        return 0
+    else:
+        return 1 + tree.left.depth
```

Comment:
I'd add a comment here mentioning that we recur on `tree.left` because (1) we're guaranteed for the `.left` attribute to be non-nil due to the tree invariant and (2) we only need to check one side since Merkel trees are "symmetric" in that all paths to leaves have the same depth.